src/meta/api/src/schema_api_impl.rs

```diff
@@ -2855,3 +2909,83 @@ async fn list_tables_from_share_db(
     )
     .await
 }
+
+async fn build_upsert_table_copied_file_info_conditions<T>(
+    this: &T,
+    req: &UpsertTableCopiedFileReq,
+    file_name_keys: &[String],
+    tb_meta_seq: u64,
+    fail_if_duplicated: bool,
+) -> Result<(Vec<TxnCondition>, Vec<TxnOp>), KVAppError>
+where
+    T: kvapi::KVApi<Error = MetaError>,
+{
+    let table_id = req.table_id;
+    let tbid = TableId { table_id };
+
+    let mut condition = vec![txn_cond_seq(&tbid, Eq, tb_meta_seq)];
+    let mut if_then = vec![];
+
+    // TODO put this doc in upsert_table_copied_file_info?
+    // `remove_table_copied_files` and `upsert_table_copied_file_info`
+    // all modify `TableCopiedFileInfo`,
+    // so there used to has `TableCopiedFileLockKey` in these two functions
+    // to protect TableCopiedFileInfo modification.
+    // In issue: https://github.com/datafuselabs/databend/issues/8897,
+    // there is chance that if copy files concurrently, `upsert_table_copied_file_info`
+    // may return `TxnRetryMaxTimes`.
+    // So now, in case that `TableCopiedFileInfo` has expire time, remove `TableCopiedFileLockKey`
+    // in each function. In this case there is chance that some `TableCopiedFileInfo` may not be
+    // removed in `remove_table_copied_files`, but these data can be purged in case of expire time.
+
+    let mut file_name_infos = req.file_info.clone().into_iter();
+
+    if fail_if_duplicated {
+        // "fail_if_duplicated" mode, assumes files are absent
+        for (file_name, file_info) in file_name_infos {
```

Comment:
maybe we can optimize here:

before copy,  we got a table version with a list  of files to copy, these files not exists as long as  table version  not changed, so no need to check their existence if  table version not changed since then.

Reply:
hope I get it right, seems that is exactly what this PR does (in non-force mode):

in one kv tx,  we assume that

- the table version does not change
- the files are absent
   these files are collected before constructing the kv tx

~~~
    if  table_id.version = expected_table_version
          // file_copied_1_key is blended with table id
         // version eq 0, indicates that this file is absent
         && file_copied_1_key.version == 0 
         && file_copied_2_key.version == 0
         ...
          && file_copied_n_key.version == 0
     then 
          put_table_meta(...);
          put(file_copied_1_key, file_copied_1_info);
          .....
          put(file_copied_1_key, file_copiedn1_info);
     else 
         get(table_id)
~~~         

if the above tx failed (with a response, that contains the output of `else branch`, i.e. the

~~~
     else 
         get(table_id)
~~~

we check the table meta returned (note , in this case , the kv tx failed):

if the table version is not changed, according to the output of kv tx, then during the executions of this kv tx, there must be some duplicated files. then 
 - if not in `force` mode, we just return with the error `DuplicatedUpsertFiles` **immediately**, without retrying.
 - if in `force` mode, we fallback to the previous logic (before this PR), that will retry

Reply:
if in `force mode`, the files could be put "unconditionally" (except that the table id should exist, and the table version match),  the `if condition` can be adjusted to 

~~~
    if  table_id.version = expected_table_version   // (without check the field are absent)
~~~

but again, I am not sure if this may cause troubles (other components that depend on the current logic of implementation)